src/context.rs

```diff
@@ -0,0 +1,35 @@
+//! All state shared by the lexer, parser and semantic analysis phases.
+use crate::{asciifile::AsciiFile, diagnostics::Diagnostics, lexer::Spanned, strtab::StringTable};
+use failure::AsFail;
+use termcolor::{ColorChoice, StandardStream, WriteColor};
+
+pub struct Context<'ctx> {
+    pub file: AsciiFile<'ctx>,
+    pub strtab: StringTable,
+    pub diagnostics: Diagnostics,
+}
+
+impl<'ctx> Context<'ctx> {
+    pub fn new(file: AsciiFile<'ctx>, writer: Box<dyn WriteColor>) -> Self {
+        Self {
+            file,
+            strtab: StringTable::new(),
+            diagnostics: Diagnostics::new(writer),
+        }
+    }
+
+    pub fn default(file: AsciiFile<'ctx>) -> Self {
```

Comment:
The exercises require a particular output behavior. Although it is limited to `stdout`, and this default is for `stderr`, I would rather have `write` injected for any instantiation of `Context`.
Although it's unlikely that there will be a requirement on `stderr` output format, I would rather have an overview of where `stderr` is written to right from `main.rs:run_compiler()`